File: python/d11.py
from collections import defaultdict, deque
from functools import cache
from itertools import pairwise
from math import prod

from utils import file_path

with open(file_path(day=11), "r") as file:
    connections = {
        from_node[:-1]: to_nodes
        for from_node, *to_nodes in [
            line.strip().split(" ") for line in file.readlines()
        ]
    }


@cache
def dfs(node: str) -> int:
    if node == "out":
        return 1
    return sum(dfs(n) for n in connections[node])


print("p1", dfs("you"))


def topo_sort() -> list[str]:
    L = deque([])
    nodes = set(connections.keys())
    marked = set()

    def visit(node: str):
        if node in marked:
            return
        for n in connections.get(node, []):
            visit(n)
        marked.add(node)
        L.appendleft(node)

    while nodes - marked:
        node = (nodes - marked).pop()
        visit(node)

    return list(L)


def count_paths_dag(order, from_node, to_node):
    dp = defaultdict(int)
    dp[from_node] = 1
    for node in order:
        for n in connections.get(node, []):
            dp[n] += dp[node]
    return dp[to_node]


order = topo_sort()

print(
    "p2",
    sum(
        prod(count_paths_dag(order, s, t) for s, t in pairwise(path))
        for path in [
            ("svr", "fft", "dac", "out"),
            ("svr", "dac", "fft", "out"),
        ]
    ),
)

File: rust/src/bin/d11.rs
use std::collections::HashMap;

use aoc25::day_input;
use itertools::Itertools;

fn main() {
    let input = day_input!("11");
    let graph = parse_input(input);
    println!("part one: {}", part_one(&graph));
    println!("part two: {}", part_two(&graph));
}

type Node = &'static str;
type Graph = HashMap<Node, Vec<Node>>;
type Cache = HashMap<(Node, Node), usize>;

fn parse_input(input: &'static str) -> Graph {
    input
        .lines()
        .map(|line| {
            let mut parts = line.split(' ');
            let to = parts.next().unwrap().strip_suffix(':').unwrap();
            (to, parts.collect())
        })
        .chain(vec![("out", vec![])])
        .collect()
}

fn n_paths(graph: &Graph, node: &'static str, target: &'static str, cache: &mut Cache) -> usize {
    if cache.contains_key(&(node, target)) {
        return *cache.get(&(node, target)).unwrap();
    }

    let result = (node == target) as usize
        + graph[node]
            .iter()
            .map(|n| n_paths(graph, n, target, cache))
            .sum::<usize>();

    cache.insert((node, target), result);
    result
}

fn part_one(graph: &Graph) -> usize {
    n_paths(graph, "you", "out", &mut HashMap::new())
}

fn part_two(graph: &Graph) -> usize {
    let paths = [["svr", "fft", "dac", "out"], ["svr", "dac", "fft", "out"]];
    paths
        .iter()
        .map(|path| {
            path.iter()
                .tuple_windows()
                .map(|(a, b)| n_paths(graph, a, b, &mut HashMap::new()))
                .product::<usize>()
        })
        .sum()
}
